src/balance.rs

```diff
@@ -26,41 +41,219 @@ impl<T: Config> From<Error<T>> for StatusCode {
+        }
+    }
+
+    /// Does a state checking on initial state of balances with current state.
+    fn cmp_with_initial_state(&self) -> DispatchResult {
+        use core::cell::Ref;
+
+        use sp_runtime::traits::Zero;
+
+        let init_state: Ref<HashMap<AccountIdOf<T>, BalanceOf<T>>> = self.initial_state.borrow();
+        let balances: Ref<HashMap<AccountIdOf<T>, BalanceOf<T>>> = self.balances.borrow();
+        if init_state.len() != balances.len() {
+            return Err(DispatchError::Corruption);
+        }
+
+        let mut sum_initial = BalanceOf::<T>::zero();
+        let mut sum_balances = BalanceOf::<T>::zero();
+        init_state.values().for_each(|b| sum_initial += *b);
+        balances.values().for_each(|b| sum_balances += *b);
+
+        if sum_initial == sum_balances {
+            Ok(())
+        } else {
+            Err(DispatchError::Corruption)
         }
     }
 }
 
-impl<T: Config> Default for BalanceAdapter<T> {
+impl<T: Config> Default for BalanceAdapter<T>
+where
+    AccountIdOf<T>: std::cmp::Eq + std::hash::Hash,
+    BalanceOf<T>: From<u128> + Into<u128>,
+{
     fn default() -> Self {
         Self::new()
     }
 }
 
-impl<T: Config> BalanceHandler for BalanceAdapter<T> {
+impl<T: Config> BalanceHandler for BalanceAdapter<T>
+where
+    AccountIdOf<T>: std::cmp::Eq + std::hash::Hash,
+    BalanceOf<T>: From<u128> + Into<u128>,
+{
     type Error = StatusCode;
 
     fn transfer(
         &self,
-        _src: AccountAddress,
-        _dst: AccountAddress,
-        _cheque_amount: u128,
+        src: AccountAddress,
+        dst: AccountAddress,
+        cheque_amount: u128,
     ) -> Result<bool, Self::Error> {
         // TODO(rqnsom): This won't break the MoveVM unless the balance module is accessed.
-        unimplemented!()
+
+        // use move_core_types::vm_status::StatusCode::{ARITHMETIC_ERROR, INSUFFICIENT_BALANCE};
+        use move_core_types::vm_status::StatusCode::INSUFFICIENT_BALANCE;
+
+        let from = Pallet::<T>::to_native_account(&src)?;
+        let to = Pallet::<T>::to_native_account(&dst)?;
+        let amount = BalanceOf::<T>::from(cheque_amount);
+
+        self.ensure_account_exists(&from);
+        self.ensure_account_exists(&to);
+
+        // if *(self.balances.borrow().get(&from).ok_or(ARITHMETIC_ERROR)?) > amount {
+        //     {
+        //         let mut balances = self.balances.borrow_mut();
+        //         *(balances.get_mut(&from).ok_or(ARITHMETIC_ERROR)?) -= amount;
+        //     }
+        //     if let Some(balance_to) = self.balances.borrow_mut().get_mut(&to) {
+        //         *balance_to += amount;
+        //         Ok(true)
+        //     } else {
+        //         let mut balances = self.balances.borrow_mut();
+        //         *(balances.get_mut(&from).ok_or(ARITHMETIC_ERROR)?) += amount;
+        //         Err(ARITHMETIC_ERROR)
+        //     }
+        // } else {
+        //     Err(INSUFFICIENT_BALANCE)
+        // }
```

Comment:
Let's remove this

Reply:
You are sure, that my evil-`unwrap`s are ok? :wink: